### ISSUE TITLE
Java crashes due to hsperfdata file conflicts across sandboxes

### ISSUE THREAD

**Issue by or-shachar** [bazelbuild/bazel#3236]
Running `bazel build` on a fat java/scala project (several thousands of targets) fails when working on linux debian with user namespace enabled.

### Issue
Trying to run `bazel build` with user namespace enabled:
```
$ sysctl kernel.unprivileged_userns_clone=1
```
The build runs alright but at some point it crashes with weird memory issue:
```
ERROR: <target-path>/BUILD:35:1: error executing shell command: '
  rm -rf bazel-out/local-fastbuild/bin/<package>/<target>.jar_temp_resources_dir
  set -e
  mkdir -p bazel-out/local-fastbuild/bin/<target>' failed: Process terminated by signal 6 [sandboxed].
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGBUS (0x7) at pc=0x00007f094606874b, pid=5, tid=0x00007f09472e0700
#
# JRE version:  (8.0_131-b11) (build )
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.131-b11 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# V  [libjvm.so+0x96874b]  PerfMemory::alloc(unsigned long)+0x7b
#
# Core dump written. Default location: /home/builduser/.cache/bazel/_bazel_builduser/bc0e462ab01ac9379d22ad058ca1cb1f/bazel-sandbox/4864102460254154064/execroot/__main__/core or core.5
#
# An error report file with more information is saved as:
# /home/builduser/.cache/bazel/_bazel_builduser/bc0e462ab01ac9379d22ad058ca1cb1f/bazel-sandbox/4864102460254154064/execroot/__main__/hs_err_pid5.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#
```

### Environment info
The machine is docker container based on debian image
```
$ uname -a
Linux 167-docker99 3.16.0-4-amd64 #1 SMP Debian 3.16.43-2 (2017-04-30) x86_64 GNU/Linux
builduser@167-docker99:~/ws/bazel-port-isolation$ cat /etc/*-release
PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
NAME="Debian GNU/Linux"
VERSION_ID="8"
VERSION="8 (jessie)"
ID=debian
HOME_URL="http://www.debian.org/"
SUPPORT_URL="http://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"Bazel version
```

- The project is fat (several thousands of **java** / **scala** targets)
- Bazel was built from a81264e1043dd90e984d9fcef5ce9962dce90d1d
- **rules_scala** - https://github.com/wix/rules_scala/commit/d66c9d7506ecc6a1b454055b35bd10ef064b9d98 (basically https://github.com/bazelbuild/rules_scala/commit/5d6ff512652b8b55f5d26f6ea69e05d86582d996 with small changes around `specs2` versions and test runner env preparation)

### additional information
- issue does not happen when `unprivileged_userns_clone=0` (but clearly - that's not a solution)
- with user namespace enabled, bazel [0.5.1](https://github.com/bazelbuild/bazel/releases/tag/0.5.1) showed [this issue](https://stackoverflow.com/questions/44656481/port-isolation-with-bazel-inside-docker) . May also be related to #3064 .

**Comment by philwo** [comment]
Can you please send the output of "cat /proc/mounts", "free -m", "df -h" and "df -i" from inside the Docker container where the SIGBUS happened?

I think this happens when /tmp does not have enough space or is a tmpfs and there's not enough free RAM to back all files that the JVM wants to create via mmap there, but I had a hard time tracking it down exactly.

**Comment by or-shachar** [comment]
The machine should be crazy strong...  Maybe even too strong? :-)
```
#cat /proc/mounts
rootfs / rootfs rw 0 0
none / aufs rw,relatime,si=bd2d63d83e471179,dio,dirperm1 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev tmpfs rw,nosuid,mode=755 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=666 0 0
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /sys/fs/cgroup tmpfs rw,nosuid,nodev,noexec,relatime,mode=755 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/lib/systemd/systemd-cgroups-agent,name=systemd 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
mqueue /dev/mqueue mqueue rw,nosuid,nodev,noexec,relatime 0 0
/dev/xvda2 /etc/resolv.conf ext4 rw,relatime,data=ordered 0 0
/dev/xvda2 /mnt/tmp ext4 rw,relatime,data=ordered 0 0
/dev/xvda2 /tmp/config ext4 rw,relatime,data=ordered 0 0
/dev/xvda2 /opt/intel ext4 rw,relatime,data=ordered 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev 0 0
/dev/xvda2 /opt/bin ext4 rw,relatime,data=ordered 0 0
/dev/xvda2 /etc/hostname ext4 rw,relatime,data=ordered 0 0
/dev/xvda2 /opt/private ext4 rw,relatime,data=ordered 0 0
/dev/xvda2 /etc/hosts ext4 rw,relatime,data=ordered 0 0
/dev/xvda2 /home/builduser/.npmrc ext4 ro,relatime,data=ordered 0 0
tmpfs /run/docker.sock tmpfs rw,nosuid,relatime,size=52934720k,mode=755 0 0
/dev/xvda2 /home/builduser/.pypirc ext4 ro,relatime,data=ordered 0 0
/dev/xvda2 /home/builduser/.ssh/google_compute_engine ext4 rw,relatime,data=ordered 0 0
/dev/xvda2 /home/builduser/.jfrog/jfrog-cli.conf ext4 ro,relatime,data=ordered 0 0
/dev/xvda2 /home/builduser/.ssh/google_compute_engine.pub ext4 rw,relatime,data=ordered 0 0
devpts /dev/console devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0

#free -m
             total       used       free     shared    buffers     cached
Mem:        258470      43545     214924         32       4771      28253
-/+ buffers/cache:      10520     247950
Swap:            0          0          0

# df -h
Filesystem      Size  Used Avail Use% Mounted on
none             99G   36G   60G  38% /
tmpfs           127G     0  127G   0% /dev
tmpfs           127G     0  127G   0% /sys/fs/cgroup
/dev/xvda2       99G   36G   60G  38% /mnt/tmp
tmpfs           127G   16M  127G   1% /dev/shm
tmpfs            51G   17M   51G   1% /run/docker.sock

# df -i
Filesystem       Inodes   IUsed    IFree IUse% Mounted on
none            6553600 1186838  5366762   19% /
tmpfs          33084198     135 33084063    1% /dev
tmpfs          33084198      14 33084184    1% /sys/fs/cgroup
/dev/xvda2      6553600 1186838  5366762   19% /mnt/tmp
tmpfs          33084198      16 33084182    1% /dev/shm
tmpfs          33084198     966 33083232    1% /run/docker.sock
```

**Comment by philwo** [comment]
Whoah, yes, resources are not an issue on that machine. :)

Still, the only thing I could ever find about JVMs crashing with SIGBUS is this: http://bugs.java.com/view_bug.do?bug_id=6563308

I wonder if it's the same issue. Do you use any --sandbox_* flags for your build? Maybe --sandbox_tmpfs_path=...?

**Comment by or-shachar** [comment]
No... I have some *test* targets where I set the `java_flags` with `-Djava.io.tmpdir=/tmp` but it fails on the build phase.

**Comment by ittaiz** [comment]
@philwo any ideas? the problem is that network-isolation on our machine (docker container) doesn't work without enabling user namespace, fails for a different reason on 0.5.1 when it's enabled and doesn't build on head when it's enabled.
We were able to verify on a very small java repo that on HEAD with user namespace enabled that network-isolation works for us.
If we could just get past this we could parallel our tests...

**Comment by philwo** [comment]
I'm trying to find out why this is happening now.

**Comment by philwo** [comment]
Can you please try using the flag "--sandbox_tmpfs_path=/tmp" in your "bazel build" command and see if the error still happens? This will mount an empty tmpfs on /tmp for each running action.

An alternate idea might be to mount a tmpfs on /tmp inside the Docker container before running the "bazel build" command, but I'd like to try the first one first, because I remember that it helped a different user.

It would be interesting to see if this makes the problem disappear.

**Comment by philwo** [comment]
A process gets SIGBUS in one of two conditions:

1) It tries to read an address from a file via mmap() that no longer exists (e.g. because the file was truncated by a different process).
2) It tries to write more bytes to an mmap'd file than the underlying device can hold.

The first thing might happen when multiple JVM processes try to use the same file via mmap and one of them truncates it, I guess. Is the failing action always the same kind, e.g. a Scala compilation?

The second possibility is more likely, but you have 60G free on / and /tmp is not a mount point, so it's hard to see how you can run out of disk space during a build... OTOH you have almost four times as much RAM as free disk space on that machine - maybe some process uses a heuristic like "let's allocate a temp file with the size of 1/4 the RAM, this should always be a reasonable number"?

I also don't understand why this only happens when you use the linux-sandbox and not when you use the standalone strategy. Do I understand correctly that the build works fine then?

**Comment by ittaiz** [comment]
@philwo using `--sandbox_tmpfs_path=/tmp` indeed passes the build, thanks!
It seems that it slows down the build a bit. Am I right? Didn't have time yet to run multiple benchmarks.

**Comment by ittaiz** [comment]
Also, this should be solved right? Shouldn't it be a release blocker?

**Comment by philwo** [comment]
> using --sandbox_tmpfs_path=/tmp indeed passes the build, thanks!

Nice! :)

> It seems that it slows down the build a bit.

It shouldn't.. actually it should make things a bit faster, because /tmp is now backed by RAM instead of disk. On the other hand, it might be possible that mounting the tmpfs incurs some overhead, too.

If you measure a noticeable difference, I'd be quite interested in it.

> this should be solved right? Shouldn't it be a release blocker?

Absolutely! The problem is, I cannot reproduce this on my machine and the cause is completely unknown. :( If we had a clear repro case or someone more familiar with JVM internals, it might be easier to get to the bottom of this issue.

**Comment by philwo** [comment]
What I'd be really interested in is if simply mounting a tmpfs on /tmp also helps, or if the issue then happens again. What I mean is:

```
# I'm in my Docker container!
$ mount -t tmpfs tmpfs /tmp
$ bazel build //...
```

If the issue reoccurs, I believe that we're seeing a race condition here, maybe related to this: https://stackoverflow.com/questions/76327/how-can-i-prevent-java-from-creating-hsperfdata-files

Maybe the JVMs in a highly parallel build accidentally create hsperfdata files with the same name and when one truncates the file of another running JVM, that JVM gets a SIGBUS because the file underlying its mmap went away. But this is really just a guess.

**Comment by philwo** [comment]
OMG, wait, I got it

**Comment by philwo** [comment]
@aehlig @ulfjack FYI.

root@ubuntu:~# strace -f -- java HelloWorld
[pid  1432] open("/tmp/hsperfdata_root", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 4
[pid  1432] fchdir(4)                   = 0
[pid  1432] open("1431", O_RDWR|O_CREAT|O_NOFOLLOW, 0600) = 5
[pid  1432] ftruncate(5, 0)             = 0
[pid  1432] mmap(NULL, 32768, PROT_READ|PROT_WRITE, MAP_SHARED, 5, 0) = 0x7f54cad93000

Every JVM creates a temporary performance instrumentation file in /tmp/hsperfdata_$USERNAME/$PID. When we use sandboxing, we use PID namespaces, which means that the PIDs are virtualized and all running JVMs believe they are PID 2.

This means that they all open/ftruncate/mmap the same file and that gives you SIGBUS eventually, due to case 1 I mentioned above: "It tries to read an address that no longer exists from an mmap'd file".

When you use --sandbox_tmpfs_path=/tmp, each running sandbox gets its own /tmp, so the files don't conflict.

This means the solution is quite simple and I'll come up with something on Monday. For now, I'd recommend to use the --sandbox_tmpfs_path=/tmp flag.

**Comment by ittaiz** [comment]
Well done!

**Comment by ulfjack** [comment]
Does it help if we set TMPDIR to a unique path for each action?

**Comment by ittaiz** [comment]
Is that targeted to me? If so how can I do that?

**Comment by philwo** [comment]
I don't think there is any way to change this. "/tmp" seems to be the hard-coded location.

```bash
philwo@ubuntu:~$ strace -f \
  -E TMP=/home/philwo/tmp \
  -E TMPDIR=/home/philwo/tmp -- \
  java -Djava.io.tmpdir=/home/philwo/tmp HelloWorld 2>&1 | \
  fgrep hsperfdata
[pid  1589] open("/tmp/hsperfdata_philwo", O_RDONLY|O_NOFOLLOW) = 3
[pid  1589] open("/tmp/hsperfdata_philwo", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 4
[pid  1589] mkdir("/tmp/hsperfdata_philwo", 0755) = -1 EEXIST (File exists)
[pid  1589] lstat("/tmp/hsperfdata_philwo", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
[pid  1589] open("/tmp/hsperfdata_philwo", O_RDONLY|O_NOFOLLOW) = 3
[pid  1589] open("/tmp/hsperfdata_philwo", O_RDONLY|O_NONBLOCK|O_DIRECTORY|O_CLOEXEC) = 4
[pid  1589] unlink("/tmp/hsperfdata_philwo/1588") = 0
```

This is consistent with what I found on the Java bugtracker:
- Initial bug report: http://bugs.java.com/view_bug.do?bug_id=6447182
- Fix: http://bugs.java.com/view_bug.do?bug_id=6938627
- Rollback: http://bugs.java.com/view_bug.do?bug_id=7009828

No further action or intend to change this behavior since then.

**Comment by philwo** [comment]
Some ideas:

1) Always mount tmpfs on /tmp. Problem: Some users might not want this behavior, e.g. on memory constrained environments or if they store huge temporary files in /tmp.
2) Make /tmp/hsperfdata_$USERNAME read-only inside the sandbox. This will prevent the JVM from creating the conflicting file. Problem: Adds JVM specific knowledge to our sandbox implementation.
3) Use -XX:-UsePerfData or -XX:+PerfDisableSharedMem flag. Problem: Unclear how we can make sure that this will always be used, even when users manually launch Java from their own Skylark actions or something.

Any other ideas? I think I like 2) the best and it would be simple to implement.

**Comment by jmmv** [comment]
I don't know what the constraints are here, but instead of mounting a tmpfs instance, could you mount --bind a fresh temporary directory onto /tmp? Should avoid the memory issues, though you'd still be paying the cost to mount the extra fs. (Given that this is a problem caused by PID namespaces, which are Linux-specific, I'm assuming this unportable solution would be acceptable.)

**Comment by philwo** [comment]
@jmmv In theory yes, but this will unfortunately break the people who put their workspace or output base inside /tmp again. I still have no clue why anyone would do that, but it comes up every single time I accidentally break it.

**Comment by ulfjack** [comment]
We could detect that case and construct a sequence of bind mounts that make the workspace / output base visible even though we're bind mounting an empty dir to /tmp. Alternatively, we could bind mount an empty directory to /tmp/hsperfdata_<username>/.

**Comment by philwo** [comment]
I'll give the "empty dir on /tmp" idea a try today and if that doesn't work out go for the make "/tmp/hsperfdata_$USERNAME read-only in sandboxes" version.

**Comment by ittaiz** [comment]
Any update?

**Comment by philwo** [comment]
@ittaiz Unfortunately I got sick just after writing that comment and was out of office the entire week :| I'm fine again now and will be back on Monday. Current plan is to mkdir that directory and then make it read-only unless a tmpfs is mounted on /tmp (because that also solves the problem in a different way).

It should be a rather simple fix that I can get easily done on Monday.

**Comment by ittaiz** [comment]
Thanks! Glad to hear you're better.

**Comment by philwo** [comment]
I was just reminded by a colleague that this is still open - I'll try to fix it (once and for all!) tomorrow.

**Comment by or-shachar** [comment]
Hey @philwo ! Just confirmed that the issue persists on [0.5.3](https://github.com/bazelbuild/bazel/tree/0.5.3) (still requires the workaround).

Do we have plans to push a fix for that on 0.5.4?

Thanks!

**Comment by davido** [comment]
I found out, that performance instrumentation directory `hsperfdata_$user` created by JVM is poisoning the build on Windows, and conflicting with Bazel's `zipper` utility: #3542.

**Comment by ittaiz** [comment]
Hi,
Any news?

**Comment by ittaiz** [comment]
Hi,
Any news?

**Comment by philwo** [comment]
Hi Ittai and others,

really sorry for the lack of action here. :(

The recommended fix for this is to use --sandbox_tmpfs_path=/tmp. On top of providing a fix for this issue, it also generally improves performance and hermeticity of builds. The only known reason why someone would *not* want to use that flag is if your actions produce so much output in /tmp that it would overflow your RAM, or that your workspace or output_base is located inside /tmp (I don't understand why people would do that, but it was a common complaint when I initially made that flag the default behavior).

All the other possible fixes have some other drawbacks. That said, I will still implement a "permanent" fix for this, even for the case where --sandbox_tmpfs_path cannot be used. If this is really blocking you, please let me know and I can prioritize this.

Philipp

**Comment by hhclam** [comment]
We chatted this offline. This issue is made worse by the fact that Bazel 0.7 now ships with turbine as part of the java toolchain and running it in sandbox can cause sigbus. We talked about easy fixes to this issue such as:
1. Always put -XX:-UsePerfData in all java executions
2. In the sandbox create a symlink from hsprefdata_{user} to /dev/null when the sandbox is running.

Both will work but (2) is a bit intrusive if user has pointed hsprefdata to some other locations.

**Comment by philwo** [comment]
Fix is out: https://bazel-review.googlesource.com/c/bazel/+/23070

Does this look OK to you @ittaiz and @hhclam?

**Comment by zhongleiwang** [comment]
Hi,
will the fix be committed soon after half year? I ran into this problem pretty often recently. It will be great that a right fix can be merged soon. thanks!

**Comment by sdqali** [comment]
We are frequently encountering this issue with `0.18.0` and JDK 9. I can open a new issue if that is more helpful.

**Comment by or-shachar** [comment]
Not sure on which version is was fixed, but today we run with bazel 0.24.0 and without the `--sandbox_tmpfx_path` workaround and the build works well.
@sdqali  - you may want to open a new issue if it still happens to you

**Comment by philwo** [comment]
This is not fixed and it probably just worked due to luck, because this is a race condition. It is not safe to run Java based tools in a sandboxed Bazel build on Linux without this flag.

Sorry that it isn't fixed yet. :(

**Comment by nkoroste** [comment]
Any updates on this? I started hitting this issue very often on all linux machines in CI.

**Comment by philwo** [comment]
I'm so sorry that we don't have a good automatic fix for this yet.

@nkoroste Your method is great, I would recommend using that to everyone for now.

**Comment by fmeum** [comment]
There are more cases of files created by the JVM under hardcoded paths of the type `/tmp/<something>$PID`. A very prominent one is [`.java_pid$PID`](https://github.com/openjdk/jdk/blob/7c9868c0b3c9bd3d305e71f91596190813cdccce/src/jdk.attach/linux/classes/sun/tools/attach/VirtualMachineImpl.java#L46), which is used by the [Attach API](https://docs.oracle.com/javase/7/docs/technotes/guides/attach/index.html) to attach Java agents at runtime. Due to collisions between sandboxes, running sandboxed Java tests attaching agents in parallel will result in all kinds of non-deterministic behavior ranging from agents getting attached to the wrong test run internal crashes within the Attach API.

This does seem to be a major concern as many popular mocking frameworks for Java and Kotlin, such as [mockk](https://github.com/mockk/mockk) and [mockito](https://github.com/mockito/mockito) (the latter is used by Bazel itself), rely on this functionality through [ByteBuddy](https://github.com/raphw/byte-buddy). This has been reported in https://github.com/mockk/mockk/issues/627 and https://github.com/mockito/mockito/issues/1879, as well as https://github.com/raphw/byte-buddy/issues/1111, causing `bazel test` to fail non-deterministically. The usual workarounds are to tag tests relying on these libraries as `"exclusive"` or `"local"`, but knowing the root cause, https://github.com/bazelbuild/bazel/issues/3236#issuecomment-587752609 is certainly much better.

@philwo You can find a clean reproducer at https://github.com/raphw/byte-buddy/issues/1111 (setting `--runs_per_test=20` has always produced failed runs for me, even though this is of course non-deterministic). This basically breaks using any JVM mocking libraries in Bazel with default settings. Should this be raised in priority given that it affects very realistic and simple tests? Even if fixing this by default turns out to be impossible without breaking other use cases, perhaps a descriptive error message linking to this issue could be added to the sandbox when this particular socket is `connect`ed to from within the sandbox.

**Comment by aryeh-looker** [comment]
+1

**Comment by lberki** [comment]
@larsrc-google I think I just ran into this (I Cced you the internal bug)

**Comment by fmeum** [comment]
@lberki @larsrc-google Happy to discuss possible solutions to this. I still think it's quite important to fix.

**Comment by larsrc-google** [comment]
Setting `--sandbox_tmpfs_path=/tmp` is the only recommendation I have. Doing that internally ran into one issue (TMPDIR set to a subdirectory of /tmp) that had us roll it back. I'm thinking of an "auto" value for `--sandbox_tmpfs_path` that would first look for TMPDIR and friends before using a default `/tmp`.

**Comment by fmeum** [comment]
> Setting `--sandbox_tmpfs_path=/tmp` is the only recommendation I have. Doing that internally ran into one issue (TMPDIR set to a subdirectory of /tmp) that had us roll it back. I'm thinking of an "auto" value for `--sandbox_tmpfs_path` that would first look for TMPDIR and friends before using a default `/tmp`.

The main problem I see with some version of a tmpfs being the default is that some tests can produce GBs of temporary file and would quickly exhaust the available RAM. 

@larsrc-google What do you think of having the sandbox runner create a dedicated subdirectory of `TMPDIR` on the host and mounting that into the sandbox as `/tmp`? With this behavior, different sandboxes can no longer collide but temporary files would still consume the same resource (disk space).

**Comment by larsrc-google** [comment]
Sounds like a good approach.

**Comment by lberki** [comment]
Why do we have a PID namespace in the sandbox?

The reason why this bug exists is that actions have separate PID namespaces but a mostly shared file system, which strikes me as odd: we either try to isolate actions as fully as possible (but then how come they share `/tmp`?) or we only try to protect against mostly-accidental hermeticity violations (but then why the PID namespace?)

**Comment by philwo** [comment]
@fmeum Sounds like a good idea!

@lberki Because using a PID namespace is the fastest and most reliable way available on Linux to reliably keep track of child processes even when they try to daemonize. When PID 1 exits, the kernel immediately, automatically and 100% reliably kills and cleans up all processes that were spawned inside the sandbox.

It also increases reproducibility and hermeticity by preventing processes from looking at other running processes.

They share the host’s filesystem and /tmp, because that proved to be the correct balance between isolation and convenience for most users and also the level we could most likely consistently provide across platforms. There’s no way to isolate the action’s  filesystem from the host on macOS or Windows for example. (I guess there are mechanisms on the latter, but probably too complex and slow for our purpose.)

**Comment by lberki** [comment]
@philwo argh,  I did not think of that, thanks for the explanation. And now you have awakened in me a yearning for the process control API of Windows, which does solve this particular problem quite well; it's kinda weird on Linux that if you want to reliably kill all your descendants, you either have to put them also in a file system namespace or live with breaking the invariant that PIDs are unique across processes accessing the same file system. Grump.

As tradeoffs go, I'd take occasional JVM crashes any day over occasionally leaving daemon processes running because the first failure mode is at least obvious. But it looks like @fmeum has already implemented a reasonable workaround so I can have both my cake and eat it, too. Yay!

**Comment by fmeum** [comment]
While the JVM crashes during compile actions stand out and are quite searchable, I spent hours debugging that and why my mocking libraries ended up being attached to *another sandboxed test*. That's why I sincerely hope that we can get something like https://github.com/bazelbuild/bazel/pull/16336 to become the default.

**Comment by lberki** [comment]
Yeah, in fact, I have to stop myself from campaigning for the flag flip to go into Bazel 6.0 :)

**Comment by lberki** [comment]
@larsrc-google doesn't your comment https://github.com/bazelbuild/bazel/issues/3236#issuecomment-1256536513 contradict https://github.com/bazelbuild/bazel/issues/3236#issuecomment-311065879 which says that mounting an empty directory at `/tmp` doesn't work?

(I don't know what exact mechanism causes this fix to not work if the output base is under `/tmp`, but I'd trust @philwo over myself here)

**Comment by fmeum** [comment]
We could do https://github.com/bazelbuild/bazel/issues/3236#issuecomment-311276029, but can also just gracefully disable hermetic tmp if the output base is under it.

**Comment by philwo** [comment]
Haha, @lberki, you're right and that's the reason why I didn't just implement that :)

Putting a workspace or output base under `/tmp` really grinds my gears more than any sand from the sandbox could do. 😆 What happens is that by mounting a tmpfs or empty directory over `/tmp` prevents the symlinks from inside the sandbox from resolving to the files - because they'd point to files inside `/tmp/myworkspace/...` (for input files coming from the workspace) or `/tmp/outputbase/...` (for action inputs that are generated by other actions during the build), and that of course no longer works if you mount something over `/tmp` inside the sandbox.

The suggested special case of mounting an empty directory to `/tmp/hsperfdata_$username` then again feels like Bazel is hardcoding some weird workaround for a buggy behavior in JDK into its defaults, even though the build might not even use Java. Of course there is no reliable way to detect whether any given genrule or action "runs Java", so we can't reliably only do it for Java actions. I'm not really a fan of that either.

Mhm.

"Gracefully disabling hermetic tmp" means we have to print a warning message, otherwise people will not be aware that they're potentially shooting themselves in the foot, but that means it will result in one of two outcomes:
- People whose build doesn't use Java are annoyed at seeing that warning all the time. Making it possible to disable the warning e.g. via a flag in the ~/.bazelrc will likely mean that at some point the user *will* run a build using Java, not remember that they disabled the warning and get the mysterious crashes again.
- People whose build uses Java either ignore the warning, because "it has always worked fine" and then they'll still blame Bazel when the JVM crashes. Or they'll do the reasonable thing and stop putting their stuff into /tmp. Then it would have been better to just make it an error in the first place.

I cannot think of a single, elegant solution that solves this problem without some annoying caveats, and that's why this isn't fixed yet.

**Comment by larsrc-google** [comment]
Would it be possible to make the relevant PID in the namespace be either random or something given from the outside (e.g. action #)?

If not, I would say make it an error to use hermetic /tmp together with building from /tmp. Building from /tmp is uncommon and potentially problematic enough that making it conflict with hermetic /tmp avoids subtle weirdness and educates people.

**Comment by lberki** [comment]
@philwo thanks, now I get it; for some reason I thought that the sandbox individually bind mounts the input files instead of symlinking.

I'd take special-casing Java in the sandbox code over Java randomly crashing in actions any day; that way, at least it's only us who gert to see the ugliness.

I was wondering if we could get away with a more-complicated-than-seems-necessary solution:
* We bind mount the workspace, `bazel-out/` and `$OUTPUT_BASE/external/` to well-known locations in the sandbox (`/bazel_workspace`, `/bazel-out`, `/bazel-external` or something, doesn't really matter)
* We direct the symlinks in the sandbox to those directories and not to their "real" locations
* bind mount `/tmp` to an empty directory as above

Then the sandbox would see consistent output paths, it would work if the output base or the workspace is under `/tmp` and it wouldn't clash with anything on the "real" file system except with these `/bazel-*` paths if someone is mad enough to have those on their file system or maybe in nested Bazel invocations (but even then, one could add a unique string per action to the path)

However, IIRC runfiles trees contain *absolute* symlinks to their contents, so they would break if they are symlinked "naively". It's not an unsolvable problem because the remote execution strategy solves it, but it does require some extra thinking.

**Comment by lberki** [comment]
I spent some time trying to improve `--incompatible_sandbox_hermetic_tmp` so that it works also when the source tree or the execroot are under `/tmp`:

https://github.com/bazelbuild/bazel/pull/16674

**Comment by lberki** [comment]
Update: now running the JVM in the sandbox should be stable with the `--incompatible_hermetic_sandbox_tmp` command line option.

I'd appreciate if you gave it a try; we are planning to flip that flag eventually (right, @larsrc-google ?) and thus the more testing and in the more diverse environment, the better.

**Comment by JaredNeil** [comment]
> Interestingly, HotSpot recently fixed this problem, too: [openjdk/jdk@84f2314](https://github.com/openjdk/jdk/commit/84f23149e22561173feb0e34bca31a7345b43c89)

Unfortunately fixing the problem was accompanied by a [warning message](https://github.com/openjdk/jdk/commit/84f23149e22561173feb0e34bca31a7345b43c89#diff-7313eb3d328797a7720fa1b2b73cd159934506593443e45534baad80cb1382b7R924-R927) printed to STDOUT, which is breaking some of our build actions that write to STDOUT (https://github.com/google/google-java-format), and filling our build logs with hundreds of those warnings for all other JVM-tool actions. I'm not sure how we were not affected by the crash, but are now affected by the logging, but hopefully `--incompatible_sandbox_hermetic_tmp` will fix that new problem for us. We're still on Bazel 5.3, but we'll be sure to try this flag when we can.

**Comment by larsrc-google** [comment]
Here's a thing to try for the warning message: Add the following to the JVM options for the actions:

```
-Xlog:disable
-Xlog:all=warning:stderr:uptime,level,tags
```

**Comment by Conhokis** [comment]
Quick chime in, looking at `--incompatible_hermetic_sandbox_tmp`

This is fundamentally not a fix, it's a workaround for not using secure temp file creation. Enabling this workaround globally as done in https://github.com/bazelbuild/bazel/pull/18892 smells like a "bruteforce" attempt that could actually mask real issues in tools (such as the one in the JDK) that are not really prepared for parallelization, regardless whether you're running it in a isolated environment.
So seems to me it's indeed fixing a java specific problem, but pretending it's a new feature while in reality adding extra complexity to the users.

And of course flag is marked as incompatible since users are surely relying already on the assumption that /tmp is shared. 

So I think that the solution is definitely going in the right direction, but it has a couple issues:
1. Unlike `--sandbox_tmpfs_path`, I don't actually have the flexibility to provide extra paths. This is a bit of a smell that it's trying to solve a very specific problem here instead of providing an extra feature for isolation.
2. I strongly believe it should be opt-in, not opt-out. We give the flexibility for users to set this on their builds or fix the problems in their tools.

**Comment by fmeum** [comment]
> This is fundamentally not a fix, it's a workaround for not using secure temp file creation. Enabling this workaround globally as done in #18892 smells like a "bruteforce" attempt that could actually mask real issues in tools (such as the one in the JDK) that are not really prepared for parallelization, regardless whether you're running it in a isolated environment.
> So seems to me it's indeed fixing a java specific problem, but pretending it's a new feature while in reality adding extra complexity to the users.

It is both a fix and a workaround:
1) The JVM used to (until https://github.com/openjdk/jdk/commit/84f23149e22561173feb0e34bca31a7345b43c89) rely on the PID to generate the hsperfdata file. I fully agree with you that this is not a good practice and it should probably use some other mechanism to prevent collisions among these files.
2) The JVM uses a shared socket in a well-known location under `/tmp` that depends on the PID to handle certain remote access features, such as agent installation (see https://github.com/bazelbuild/bazel/issues/3236#issuecomment-922458662). This seems to be fully appropriate for this use case and I don't immediately see how this could be set up in other ways - "well-known location" and "collision-free" are pretty much at odds by definition. Since this scheme only breaks if two processes share `/tmp` but not the PID namespace, not sharing `/tmp` seemed to be the only way if we want to preserve isolation.

I think that 2) by itself is a reason for having and flipping `--incompatible_hermetic_sandbox_tmp`. I searched for but couldn't find tools that have similar issues and don't rely on `/tmp`, which is why I haven't thought about making this handling more configurable. If such tools exist, this could definitely be helpful.

**Comment by Conhokis** [comment]
Agreed for point 2, still would like to see the flexibility there but I understand now better this issue is beyond the hsperfdata file only.

Also understand that from bazel maintainers POV having out of the box support for common JVMs is a must-have.

Then the note on flexibility is done, consider it a possible minor improvement for some more esoteric tools / setups.

Appreciate the swift the reply :smile:

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.0.0 RC5](https://releases.bazel.build/7.0.0/rc5/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
